internal/tracinginterceptor/interceptor.go

```diff
@@ -39,54 +45,184 @@ var (
 // HandleOneway implements interceptor.OnewayInbound
 func (m *Interceptor) HandleOneway(ctx context.Context, req *transport.Request, h transport.OnewayHandler) error {
-	// TODO: implement
-	panic("implement me")
+	parentSpanCtx, _ := m.tracer.Extract(m.propagationFormat, getPropagationCarrier(req.Headers.Items(), req.Transport))
+	extractOpenTracingSpan := &transport.ExtractOpenTracingSpan{
+		ParentSpanContext: parentSpanCtx,
+		Tracer:            m.tracer,
+		TransportName:     req.Transport,
+		StartTime:         time.Now(),
+		ExtraTags:         commonTracingTags,
+	}
+	ctx, span := extractOpenTracingSpan.Do(ctx, req)
+	defer span.Finish()
+
+	err := h.HandleOneway(ctx, req)
+	return updateSpanWithErrorDetails(span, false, nil, err)
 }
 
 // CallOneway implements interceptor.OnewayOutbound
-func (m *Interceptor) CallOneway(ctx context.Context, request *transport.Request, out transport.OnewayOutbound) (transport.Ack, error) {
-	// TODO: implement
-	panic("implement me")
+func (m *Interceptor) CallOneway(ctx context.Context, req *transport.Request, out transport.OnewayOutbound) (transport.Ack, error) {
+	createOpenTracingSpan := &transport.CreateOpenTracingSpan{
+		Tracer:        m.tracer,
+		TransportName: m.transport,
+		StartTime:     time.Now(),
+		ExtraTags:     commonTracingTags,
+	}
+	ctx, span := createOpenTracingSpan.Do(ctx, req)
+	defer span.Finish()
+
+	tracingHeaders := make(map[string]string)
+	if err := m.tracer.Inject(span.Context(), m.propagationFormat, getPropagationCarrier(tracingHeaders, m.transport)); err != nil {
+		ext.Error.Set(span, true)
+		span.LogFields(log.String("event", "error"), log.String("message", err.Error()))
+	} else {
+		for k, v := range tracingHeaders {
+			req.Headers = req.Headers.With(k, v)
+		}
+	}
+
+	ack, err := out.CallOneway(ctx, req)
+	return ack, updateSpanWithErrorDetails(span, false, nil, err)
 }
 
 // HandleStream implements interceptor.StreamInbound
 func (m *Interceptor) HandleStream(s *transport.ServerStream, h transport.StreamHandler) error {
-	// TODO: implement
 	panic("implement me")
 }
 
 // CallStream implements interceptor.StreamOutbound
 func (m *Interceptor) CallStream(ctx context.Context, req *transport.StreamRequest, out transport.StreamOutbound) (*transport.ClientStream, error) {
-	// TODO: implement
 	panic("implement me")
 }
+
+func updateSpanWithErrorDetails(
+	span opentracing.Span,
+	isApplicationError bool,
+	appErrorMeta *transport.ApplicationErrorMeta,
+	err error,
+) error {
+	if err == nil && !isApplicationError {
+		return nil
+	}
+	ext.Error.Set(span, true)
+	if status := yarpcerrors.FromError(err); status != nil {
+		errCode := status.Code()
+		span.SetTag("rpc.yarpc.status_code", int(errCode))
+		span.SetTag("error.type", errCode.String())
+		return err
+	}
+	if isApplicationError {
+		span.SetTag("error.type", "application_error")
+
+		if appErrorMeta != nil {
+			if appErrorMeta.Code != nil {
+				span.SetTag("application_error_code", int(*appErrorMeta.Code))
```

Comment:
Is it your plan to have integer code here? It would be unreadable from the UI, right?

Reply:
For standard yarpc/http/gRPC/tchannel status code, we want to use int, and we plan to centrally convert them to human readable error code during ingestion time or query time.

For this, IIUC, appErrorMeta.Code is actually a yarpc code, so can we use the same name tag key `rpc.yarpc.status_code` if we want to keep it integer.